grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -197,6 +248,22 @@ def _AddPrimitiveTypesSchemas(self) -> None:
       cast(Dict[str, Dict[str, str]], primitive_types_schemas)
     )
 
+  def _AddRDFTypesSchemas(self) -> None:
+    """Adds the OpenAPI schemas for RDF types."""
+    if self.schema_objs is None:
+      raise AssertionError("OpenAPI type schemas not initialized.")
+
+    for rdf_type_name in rdf_types_schemas:
+      old_description = rdf_types_schemas[rdf_type_name]['description']
+      new_prefix = f"RDF type is `{rdf_type_name}` and it represents"
+
+      if not old_description.startswith(new_prefix):
+        rdf_types_schemas[rdf_type_name]["description"] = (
+          f"{new_prefix} {old_description}"
+        )
```

Comment:
Modifying a global variable is never a good idea—here it is even worse, because you get incorrect descriptions if someone instantiates `ApiGetOpenApiDescriptionHandler` multiple times.

Reply:
I agree, it's not a clean solution, but there are some details about this. Initially I added it, as you pointed out, in `_GetDescribedSchema`, but the problem is that that function is called only for parameters and fields, whereas this description is strictly type related, not field/parmeter related (compared to `protobuf.oneof` descriptions) and this would cause that the `description` field in the definitions of the schemas of the types in the `components` field of the resulting JSON to be different from what is displayed for fields, altough the description is related to the type. The verification for the prefix should solve the problem about multiple instantiations (an alternative would be to use a global boolean flag). But if there are multiple threads, there will be a race condition. Will there be multiple threads accessing the same memory?

On the other hand, having different descriptions for the component definition (which would lack the prefix) and for field/parameter definitions is not necessarily a bad thing since the OpenAPI component definition includes the name of the type.

Please let me know if I should move forward with different descriptions for the component definition (the actual definition of the type's schema) and the fields/parameters (where a reference to the type's schema is used and the description is overriden).

Reply:
> I agree, it's not a clean solution. (...)

There are two jarring problems with the current approach.

The first one is modifying a global constant. A global constant should be a global constant, it should never change. It is problematic not only because of multi-threading issues, but also because some parts of code might make assumptions about it and the bugs are then very hard to track.

The other one is mixing the data and presentation. Here, your data is the type, the format and a short description. Combining the three to present to the user is not something that should be ever persisted. When needed, you should take the data, generate a user-facing string, give it to the user and forget it. You have all the data you need to re-create it if needed, don't waste the memory. Imagine that you would like to add internationalization to your code. If you mix data and presentation, it would be just impossible.

> But if there are multiple threads, there will be a race condition. Will there be multiple threads accessing the same memory?

It's a wrong question. Requirements change. Maybe it isn't now (but maybe it is, request handlers is something that I imagine is the most basic parallelization unit for web servers), but what if it will become in the future? I doubt anyone will remember that there is some piece of code that assumed that there is no multi-threading and it should be adjusted now.

> On the other hand, having different descriptions for the component definition (which would lack the prefix) and for field/parameter definitions is not necessarily a bad thing

Then let's do it for now and adjust if there is a need for it.

Reply:
Thank you for the clear explanation! I made the change.